internal/moreos/moreos_test.go

```diff
@@ -155,6 +157,68 @@ func Test_EnsureProcessDone(t *testing.T) {
 	require.NoError(t, EnsureProcessDone(cmd.Process))
 }
 
+func Test_EnsureChildProcessDone(t *testing.T) {
```

Comment:
@codefromthecrypt could please help to check if the scenario is correct? Thank you!

Reply:
The short of this one is I would expect the test function here to directly invoke fake_func-e (not intermediated though a `go run`). and fake_func-e would invoke fake_envoy if it could or worst case `cat`. Then I would expect this test case to wait for the fake_envoy to be ready (this part is easier than using cat as you can re-use our other code and get to a known position in the process). At that point, the test would grab the process tree of fake_func-e which should only have 2 processes itself and fake_envoy.  The test would then kill -9 its child (fake_func-e) and then with some assert eventually stuff ensure the corresponding fake_envoy died, too, by using the pid it had stored earlier.

I would not necessarily expect the cmd used to run fake_func-e to use the same process attributes as the `go test` process running it shouldn't have to be kill -9'ed, and it if did and it leaked a fake_func-e, that's sortof ok. The main thing is to only use the code under test (moreos.XXX stuff) where we are simulating its use.

We have to use compiled code because we can't really do a kill -9 test any other way at least not that I know of. I have a feeling when all this is done it really won't look complicated.

Reply:
Ps I keep saying fake envoy, but I sometimes wonder if I should be saying `RequireFakeEnvoy` Have you seen it?

We already do this and handle compilation etc in a way the result is a bin we start directly. Concretely, I would expect this change to end up with a `RequireFakeFuncE` which re-uses almost everything, just a different source file. That way maintenance is also easier to reason with as how we deal with fake binaries is consistent